source/MaterialXContrib/MaterialXNode/MaterialXData.cpp

```diff
@@ -96,64 +103,124 @@ void MaterialXData::generateXML()
     _genContext.registerSourceCodeSearchPath(_librarySearchPath);
     if (shaderRef)
     {
-        _genContext.getOptions().hwSpecularEnvironmentMethod = MaterialX::SPECULAR_ENVIRONMENT_FIS;
+        _genContext.getOptions().hwSpecularEnvironmentMethod = mx::SPECULAR_ENVIRONMENT_FIS;
     }
     else
     {
-        _genContext.getOptions().hwSpecularEnvironmentMethod = MaterialX::SPECULAR_ENVIRONMENT_NONE;
+        _genContext.getOptions().hwSpecularEnvironmentMethod = mx::SPECULAR_ENVIRONMENT_NONE;
     }
     _genContext.getOptions().hwMaxActiveLightSources = 0;
     // For Maya we need to insert a V-flip fragment
     _genContext.getOptions().fileTextureVerticalFlip = true;
-    // We do not reuqire vertex shader output to XML    
-    _xmlFragmentWrapper.setOutputVertexShader(false);
-
-    // Generator XML wrapper
-    // TODO: Determine what is a suitable unique name for VP2 fragments using
-    // this as a starting identifier.
-    const std::string shaderName(_element->getName());
-    _xmlFragmentWrapper.generate(shaderName, _element, _genContext);
-}
 
-// TODO: This does not belong here. To migrate out to another class.
-void MaterialXData::registerFragments(const std::string& ogsXmlPath)
-{
-    // Register fragments with the manager if needed
-    //	
-    if (MHWRender::MRenderer* theRenderer = MHWRender::MRenderer::theRenderer())
+    // Generate shader and put into XML wrapper.
+    // TODO: Generate unique fragment names
+    _fragmentName = _element->getNamePath();
+    _fragmentName = mx::createValidName(_fragmentName);
+    mx::ShaderPtr shader = _shaderGenerator->generate(_fragmentName, _element, _genContext);
+    if (shader)
     {
-        if (MHWRender::MFragmentManager* fragmentMgr = theRenderer->getFragmentManager())
+        std::ostringstream stream;
+        // Note: This name must match the the fragment name used for registration
+        // or the registration will fail.
+        _generator.generate(_fragmentName, shader.get(), nullptr, stream);
+        _fragmentWrapper = stream.str();
+        if (_fragmentWrapper.empty())
+        {
+            return;
+        }
+        else
+        {
+            // Strip out any '\r' characters.
+            _fragmentWrapper.erase(std::remove(_fragmentWrapper.begin(), _fragmentWrapper.end(), '\r'), _fragmentWrapper.end());
+        }
+
+        // Extract out the input fragment parameter names along with their
+        // associated Element paths to allow for value binding.
+        const mx::ShaderStage& ps = shader->getStage(mx::Stage::PIXEL);
+        for (auto uniformsIt : ps.getUniformBlocks())
```

Comment:
```suggestion
        for (const auto& uniformsIt : ps.getUniformBlocks())
```

Reply:
Also, the name of the variable is misleading. It's not an STL iterator and we're iterating over an STL container.